JIT support for multi-layered SCCs

ymanton opened the pull request:
This PR adds JIT support for multi-layered SCCs and cleans up some associated code.

Depends on the corresponding VM support in https://github.com/eclipse/openj9/pull/6536.

I'll rebase and squash the commits later, but for now I've left them as-is to make reviewing easier.

dsouzai commented on runtime/compiler/control/CompilationThread.cpp:
```diff
@@ -6519,7 +6508,7 @@ TR::CompilationInfoPerThreadBase::preCompilationTasks(J9VMThread * vmThread,
          !entry->isDLTCompile() &&
          !entry->_doNotUseAotCodeFromSharedCache &&
-         reloRuntime->isRomClassForMethodInSharedCache(method) &&
+         TR_J9VMBase::get(_jitConfig, vmThread)->sharedCache()->isPointerInSharedCache(J9_CLASS_FROM_METHOD(method)->romClass) &&
```
isn't `reloRuntime->isRomClassForMethodInSharedCache(method)` a much more compact query than `TR_J9VMBase::get(_jitConfig, vmThread)->sharedCache()->isPointerInSharedCache(J9_CLASS_FROM_METHOD(method)->romClass)`? Is there a reason or change in the future that required removing the `isRomClassForMethodInSharedCache` query?

dsouzai replied:
Ah does it have to do with the duplicated code that's there in both the SharedCache class and the RelocationRuntime class?

ymanton replied:
Yes, we have two queries that are essentially identical but implemented in two separate classes. It hurts maintainability and makes the code less clear. One query is more than enough. `TR_J9SharedCache` seems like a good place to have it because it's a bit out of place in `RelocationRuntime`.

I agree this line is a bit long winded, but it's not the norm; in most other cases it's shorter. I can add an overload for `J9Method` that would shorten this line and would probably be more convenient. Thoughts?

ymanton replied:
Actually, we have 3 queries that do the same thing; `CompilationInfo` has one too.

dsouzai replied:
> I agree this line is a bit long winded, but it's not the norm; in most other cases it's shorter. I can add an overload for J9Method that would shorten this line and would probably be more convenient. Thoughts?

I don't know if `J9_CLASS_FROM_METHOD(method)->romClass` --> `method` is worth the effort to make an overload. What do you think about a static method in `TR_J9SharedCache`, ie:

```
class TR_J9SharedCache : public TR_SharedCache
   {
   ...
   static bool isMethodInSharedCache(J9Method *method)
      {
      return TR_J9VMBase::get(_jitConfig, vmThread)->sharedCache()->isPointerInSharedCache(J9_CLASS_FROM_METHOD(method)->romClass)
      }
   ...
   }
```
Then the calls would just become
```
TR_J9SharedCache::isMethodInSharedCache(method)
```
And if there's a need for the query "is this class in the SCC", there could be another static method added for the class. 
```
TR_J9SharedCache:isClassInSharedCache(clazz)
```

I'm not wedded to this idea, just throwing it out there.

ymanton replied:
I considered this, but it's hiding the `TR_J9VMBase::get(_jitConfig, vmThread)`, which is an expensive way to get you hands on an FE and should really only be used if you have no other choice IMO.

Usually we have an FE handy, or a `Compilation` or `CodeGenerator` which we can get an FE from cheaply so people should prefer that. In this series of patches I've only used `::get()` in places where it was necessary. This method in particular is a mess with its FE usage and could use a serious cleanup, but I didn't go that far and just followed suit with the rest of the code to not complicate things further.

dsouzai replied:
That's fair, I'm fine leaving it as is.

dsouzai commented on runtime/compiler/env/J9SharedCache.cpp:
```diff
@@ -401,25 +406,67 @@ TR_J9SharedCache::matchRAMclassFromROMclass(J9ROMClass * clazz, TR::Compilation
 void *
 TR_J9SharedCache::pointerFromOffsetInSharedCache(void *offset)
```
I think it would be very help to have either a doc or a comment describing how the caches are being linearized, and what the "offset" means in this multi-SCC world.

ymanton replied:
Agreed. I'll add a doxygen comment somewhere appropriate.

ymanton replied:
Done, added the following class doc for `TR_J9SharedCache`:

```cpp
/**
 * \brief An interface to the VM's shared class cache.
 *
 * This class provides an interface to the VM's shared class cache as represented by the the descriptors in J9SharedClassConfig::cacheDescriptorList.
 * The cache descriptor list is a circular linked list. It is doubly linked when J9VM_OPT_MULTI_LAYER_SHARED_CLASS_CACHE is defined
 * or singly linked otherwise.
 *
 * If J9VM_OPT_MULTI_LAYER_SHARED_CLASS_CACHE is not defined, the list consists of a single element who's next pointer refers back to itself.
 * Offsets into the shared cache represent the distance from the start of the cache. Converting between pointers and offsets can be done with
 * simple pointer arithmetic.
 *
 * If J9VM_OPT_MULTI_LAYER_SHARED_CLASS_CACHE is defined, the head of the list represents the top-most (and therefore writable)
 * layer of the shared class cache, the next element represents the (N-1)th layer, and so on. The element previous to the head represents the base layer.
 * The list of cache layers can be thought of as a single logical cache starting at layer 0 and ending at layer N. Offsets into the shared cache
 * represent the distance from the start of the cache. Converting between pointers and offsets requires traversing the list starting at the base layer.
 */
```

Also added method docs for the three public methods I changed.

dsouzai commented on runtime/compiler/env/J9SharedCache.cpp:
```diff
@@ -401,25 +406,67 @@ TR_J9SharedCache::matchRAMclassFromROMclass(J9ROMClass * clazz, TR::Compilation
+         {
+         return (void *)(ptrValue - cacheStart + offsetValue);
+         }
+      offsetValue += curCache->cacheSizeBytes;
```
I'm guessing all the caches have to be the same size right?

ymanton replied:
No, they can differ in size. The code as written should handle this.

dsouzai replied:
I see, then 2 clarifications. 

1. `cacheSizeBytes` is the not the soft limit right? Also it's total size, not the amount allocated?
2. `cacheSizeBytes` can't change between runs right?

ymanton replied:
1. Yes in both cases.
2. For the lower (read-only) layers, no. If a lower layer changes size it should invalidate all subsequent layers. The VM should take care of those details.

For the current (write) layer, I'm not sure. Whatever happens today when an SCC changes size should happen here as well. @hangshao0 might be able to answer these questions with more concrete details than I can provide.

hangshao0 replied:
cacheSizeBytes is not the soft max limit. It is the total cache size and cannot change between runs.

ymanton commented:
> 5222c90 looks ok (except for the fact that the return keyword was removed from TR::CompilationInfo::isRomClassForMethodInSharedCache; however, when I grep for this method I don't see it in the final commit so I assume it was removed at some point I haven't gotten to yet.)

Good catch, the `return` was inadvertently removed but I didn't notice because I later removed the function. I'll fix it anyway since it would break `git bisect`.

ymanton commented:
Jenkins test sanity all jdk8

ymanton commented:
Jenkins test sanity all jdk8

ymanton commented:
I've added 1bd5bfae3e255b1367a24b6df485696ec67e8c2a here to see if that resolves the anonymous class loader problem without causing other issues, it sets the correct class loader prior to firing the hook. Changed to WIP to prevent accidental merging.

ymanton commented:
Jenkins test sanity all jdk8

ymanton commented:
Jenkins test sanity all jdk8

DanHeidinga commented:
> I've added 1bd5bfa here to see if that resolves the anonymous class loader problem without causing other issues, it sets the correct class loader prior to firing the hook.

The original code ensured that the on-heap and native representation of the classloader was consistent.  Once the ClassLoader object was installed in the Class, the native side always reported the same data.  1bd5bfa changes that invariant and I'm concerned about that.

Can this be handled in the JIT code by checking for anonymous classes?  There's a `J9ClassIsAnonymous` J9Class->classFlags bit that can be checked

ymanton commented:
> The original code ensured that the on-heap and native representation of the classloader was consistent. Once the ClassLoader object was installed in the Class, the native side always reported the same data. 1bd5bfa changes that invariant and I'm concerned about that.

Indeed it caused failures so I scrapped it.

> Can this be handled in the JIT code by checking for anonymous classes? There's a J9ClassIsAnonymous J9Class->classFlags bit that can be checked

I was hoping to avoid exposing this entire anonymous classes implementation detail to hook listeners, but yes we can deal with this in the JIT.

I actually pushed https://github.com/eclipse/openj9/pull/6723/commits/1c910ac021b2f7a0b8700a7338178d1e6100aa37 (prior to your comment) which uses an unconditional `classLoader = class->hostClass->classLoader`, which works for regular and anonymous classes alike because the `hostClass` member points back to the class itself for non-anonymous classes. Can we rely on `class->hostClass == class` for regular classes or should I just check `J9ClassIsAnonymous` and call it a day?

DanHeidinga commented:
> Can we rely on `class->hostClass == class` for regular classes 

yes, see
https://github.com/eclipse/openj9/blob/41b062dfd1983f421d59d0ebfffa563513625906/runtime/vm/createramclass.cpp#L2517-L2524

I was slightly concerned about an anonymous class using another anonymous class as a host class but by the time that can be done, the anonymous class's loader will be set to the original hosts and all should be good.

ymanton commented on runtime/compiler/runtime/J9Runtime.hpp:
```diff
@@ -158,7 +158,7 @@ typedef struct TR_AOTMethodHeader {
    uintptr_t compileMethodDataStartPC;
    uintptr_t compileMethodDataSize;
-   uintptr_t compileFirstClassLocation;
+   uintptr_t unused;
```
Done. https://github.com/eclipse/openj9/issues/7015.

ymanton commented:
Sanity has passed, @dsouzai can you take a look when you have a moment? After that I'll squash the commits into two commits, one to encapsulate cleanups and gardening and another to capture the new functionality.

dsouzai commented on runtime/compiler/control/CompilationThread.cpp:
```diff
@@ -94,6 +93,7 @@
 #include "env/SystemSegmentProvider.hpp"
 #include "env/DebugSegmentProvider.hpp"
 #ifdef COMPRESS_AOT_DATA
+#include "shcdatatypes.h" // For CompiledMethodWrapper
```
I think you need a `#endif` here right?

ymanton replied:
The `#endif` already exists, together with the `#ifdef` and a large section of code in between. I just migrated the include from outside the guard to inside the guard because it's only needed when `COMPRESS_AOT_DATA` is defined.

dsouzai replied:
Ooh right

dsouzai commented:
@ymanton I can't merge because of merge conflicts :)

ymanton commented:
Jenkins test sanity all jdk8

ymanton commented:
Failure is caused by existing code that relies on our old API's behaviour of returning out of bounds SCC offsets for classes not in the SCC. I got rid of that behaviour and added an assert to make sure no one asks for an offset for a class not in the SCC.

In this particular test case a class in the SCC gets redefined, which means it's no longer in the SCC, and it trips the above mentioned assert. Fixed by https://github.com/eclipse/openj9/pull/6723/commits/cbe470be38d3e65d2ce63e9fdfd72e840fadd724.

ymanton commented:
Jenkins test sanity all jdk8

dsouzai commented:
xlinux_xl build failed because of 
```
10:43:00  fatal: remote error: upload-pack: not our ref f66f4e4b50ba270d18650ec97441238dff76d519
10:43:00  fatal: The remote end hung up unexpectedly
```
which is unfortunate because that was the test that was legitimately failing before heh... we'll have to kick it off again once the current tests are done.

dsouzai commented:
Jenkins test sanity xlinuxxl jdk8

dsouzai commented:
PR Build top level job: https://ci.eclipse.org/openj9/job/PullRequest-OpenJ9/1205/ (since it's not showing up in the list of checks)
